File: server/controllers/portfolio.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transfer = exports.portfolio = void 0;
const Account_1 = __importDefault(require("../models/Account"));
const BoughtCoin_1 = __importDefault(require("../models/BoughtCoin"));
const Coin_1 = __importDefault(require("../models/Coin"));
const portfolio = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { accountId } = req.body;
    const account = yield Account_1.default.findOne({ _id: accountId });
    const boughtCoins = yield BoughtCoin_1.default.find({ owner: accountId });
    const coins = yield Coin_1.default.find({});
    if (account) {
        const portfolio = boughtCoins.map((boughtCoin) => {
            const coin = coins.find((coin) => coin.name === boughtCoin.name);
            return {
                _id: coin._id,
                name: coin.name,
                abbreviation: coin.abbreviation,
                price: coin.price,
                amount: boughtCoin.amount,
                total: boughtCoin.amount * coin.price,
                wallet: boughtCoin.wallet,
                imageUrl: coin.imageUrl
            };
        });
        res.json({
            message: 'successfully found portfolio',
            success: true,
            portfolio,
            account: {
                balances: account.balances,
            }
        });
    }
    else {
        res.status(404).json({
            message: 'cant find account'
        });
    }
});
exports.portfolio = portfolio;
const transfer = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    const { accountId, type, data } = req.body;
    const account = yield Account_1.default.findOne({ _id: accountId });
    console.log(req.body);
    if (account) {
        if (type === 'balance-to-balance') {
            const balances = account.balances;
            const { to, from, amount } = data;
            if (amount <= 0) {
                return res.status(400).json({
                    message: 'amount must be greater than 0',
                    success: false
                });
            }
            if (balances[from] >= amount) {
                const newFromBalance = balances[from] - amount;
                const newToBalance = balances[to] + amount;
                const newBalances = {
                    [to]: newToBalance,
                    [from]: newFromBalance
                };
                yield Account_1.default.findOneAndUpdate({ _id: account._id }, { balances: newBalances });
                res.json({
                    message: 'successfully transferred',
                    success: true,
                    account: {
                        balances: newBalances,
                    }
                });
            }
            else {
                res.status(400).json({
                    message: 'insufficient funds',
                    success: false
                });
            }
        }
        else if (type === 'user:account-to-user:account') {
            const { from, to, amount } = data;
            const toAccount = yield Account_1.default.findOne({ username: to });
            if (toAccount) {
                if (account.balances.bank >= amount) {
                    const newBalance = account.balances[from] - amount;
                    const newToBalance = toAccount.balances.bank + amount;
                    const newBalances = Object.assign(Object.assign({}, account.balances), { [from]: newBalance });
                    yield Account_1.default.findOneAndUpdate({ _id: account._id }, { 'balances': newBalances });
                    yield Account_1.default.findOneAndUpdate({ _id: toAccount._id }, { 'balances.bank': newToBalance });
                    res.json({
                        message: 'successfully transferred',
                        success: true,
                        account: {
                            balances: Object.assign({}, account.balances)
                        }
                    });
                }
                else {
                    res.status(400).json({
                        message: 'insufficient funds',
                        success: false
                    });
                }
            }
            else {
                res.json({
                    message: 'cant find to account',
                    success: false
                });
            }
        }
        else if (type === 'user:coin-to-user:coin') {
            const { from, to, amount } = data;
            const fromBoughtCoin = yield BoughtCoin_1.default.findOne({ owner: accountId, name: from });
            const toBoughtCoin = yield BoughtCoin_1.default.findOne({ wallet: to });
            if (fromBoughtCoin) {
                if (toBoughtCoin) {
                    const fromCoin = yield Coin_1.default.findOne({ name: fromBoughtCoin.name });
                    const toCoin = yield Coin_1.default.findOne({ name: toBoughtCoin.name });
                    if (fromCoin) {
                        if (toCoin) {
                            if (fromBoughtCoin.amount >= amount) {
                                const fromDiff = amount / fromCoin.price;
                                const toDiff = amount / toCoin.price;
                                const newFromAmount = fromBoughtCoin.amount - fromDiff;
                                const newToAmount = toBoughtCoin.amount + toDiff;
                                yield BoughtCoin_1.default.findOneAndUpdate({ _id: fromBoughtCoin._id }, { amount: newFromAmount });
                                yield BoughtCoin_1.default.findOneAndUpdate({ _id: toBoughtCoin._id }, { amount: newToAmount });
                                res.json({
                                    message: 'successfully transferred',
                                    success: true,
                                    account: {
                                        balances: account.balances,
                                    }
                                });
                            }
                            else {
                                res.status(400).json({
                                    message: 'insufficient funds',
                                    success: false
                                });
                            }
                        }
                        else {
                            res.status(404).json({
                                message: 'cant find to coin',
                                success: false
                            });
                        }
                    }
                    else {
                        res.status(404).json({
                            message: 'cant find from coin',
                            success: false
                        });
                    }
                }
                else {
                    res.status(404).json({
                        message: 'cant find to bought coin',
                        success: false
                    });
                }
            }
            else {
                res.status(404).json({
                    message: 'cant find from bought coin',
                    success: false
                });
            }
        }
        else {
            res.status(404).json({
                message: 'cant find type',
                success: false
            });
        }
    }
    else {
        res.status(404).json({
            message: 'cant find account',
            success: false
        });
    }
});
exports.transfer = transfer;
// export const giveAllBoughtCoinAWallet = async (req: Request, res: Response) => {
//   const boughtCoins = await BoughtCoin.find({});
//   boughtCoins.forEach(async (boughtCoin) => {
//     const wallet = Wallet.generate();
//     const address = wallet.getAddressString();
//     await BoughtCoin.findOneAndUpdate({ _id: boughtCoin._id }, { '$set': { wallet: address } }, {multi:true}, );
//   });
//   res.json({
//     message: 'successfully gave all bought coins a wallet',
//     success: true,
//   });
// }

File: server/routes/portfolio.py
from pydantic import BaseModel
from fastapi import APIRouter
from fastapi.responses import JSONResponse
import datetime
import uuid
import jwt

from config import db, SECRET_KEY

router = APIRouter(prefix='/api/portfolio')
account_collection = db.get_collection('accounts')
coin_collection = db.get_collection('coins')
bought_coin_collection = db.get_collection('bought-coins')
transaction_collection = db.get_collection('transactions')

def add_transaction(type: str, data: dict):
    _id = str(uuid.uuid4())
    slug = str(uuid.uuid4())[:4].upper()

    transaction_collection.insert_one({
        '_id': _id,
        'slug': slug,
        'type': type,
        'data': data,
        'createdAt': datetime.datetime.now(),
        'updatedAt': datetime.datetime.now()
    })

class Portfolio(BaseModel):
    token: str

class Transfer(BaseModel):
    token: str
    type: str
    data: dict

@router.post('/')
async def portfolio(portfolio: Portfolio):
    try:
        payload = jwt.decode(portfolio.token, SECRET_KEY, algorithms=['HS256'])
        account = account_collection.find_one({'_id': payload['_id']})

        if account:
            bought_coins = bought_coin_collection.find({'owner': account['_id']})
            portfolio_ = []

            for bought_coin in bought_coins:
                coin = coin_collection.find_one({'name': bought_coin['name']})

                portfolio_.append({
                    '_id': coin['_id'],
                    'name': coin['name'],
                    'abbreviation': coin['abbreviation'],
                    'price': coin['price'],
                    'amount': bought_coin['amount'],
                    'total': bought_coin['amount'] * coin['price'],
                    'wallet': bought_coin['wallet'],
                    'imageUrl': coin['imageUrl']
                })

            return JSONResponse(
                {
                    'message': 'successfully found portfolio', 'success': True,
                    'portfolio': portfolio_, 'account': {'balances': account['balances']}
                }, status_code=200
            )

        else:
            return JSONResponse({'message': 'cant find account', 'success': False}, status_code=404)

    except jwt.exceptions.DecodeError:
        return JSONResponse({'message': 'invalid token', 'success': False}, status_code=401)

    except jwt.exceptions.ExpiredSignatureError:
        return JSONResponse({'message': 'token expired', 'success': False}, status_code=401)

    except Exception as e:
        return JSONResponse(
            {'message': 'unknown error', 'error': str(e), 'success': False}, status_code=500
        )

@router.post('/transfer')
async def transfer(transfer: Transfer):
    try:
        payload = jwt.decode(transfer.token, SECRET_KEY, algorithms=['HS256'])
        account = account_collection.find_one({'_id': payload['_id']})

        if account:
            if transfer.type == 'balance-to-balance':
                balances = account['balances']
                to = transfer.data['to']
                from_ = transfer.data['from']
                amount = transfer.data['amount']

                if amount <= 0:
                    return JSONResponse({'message': 'amount must be greater than 0', 'success': False}, status_code=400)

                if balances[from_] < amount:
                    return JSONResponse({'message': 'insufficient funds', 'success': False}, status_code=400)

                from_balance = balances[from_] - amount
                to_balance = balances[to] + amount
                new_balances = {from_: from_balance, to: to_balance}

                account_collection.find_one_and_update({'_id': account['_id']}, {'$set': {'balances': new_balances}})
                add_transaction(type='Account Balance to Account Balance', data={
                    'from': from_, 'to': to, 'amount': amount
                })

                return JSONResponse(
                    {
                        'message': 'successfully transferred funds', 'success': True,
                        'account': {'balances': new_balances}
                    }, status_code=200
                )

            elif transfer.type == 'user:account-to-user:account':
                from_ = transfer.data['from']
                to = transfer.data['to']
                amount = transfer.data['amount']
                to_account = account_collection.find_one({'username': to})

                if amount <= 0:
                    return JSONResponse({'message': 'amount must be greater than 0', 'success': False}, status_code=400)

                if to_account:
                    if to_account['username'] == account['username']:
                        return JSONResponse({'message': 'cant transfer to yourself', 'success': False}, status_code=400)

                    if account['balances'][from_] < amount:
                        return JSONResponse({'message': 'insufficient funds', 'success': False}, status_code=400)

                    new_balance = account['balances'][from_] - amount
                    new_to_balance = to_account['balances']['bank'] + amount
                    new_balances = account['balances']
                    new_balances.update({from_: new_balance})

                    account_collection.find_one_and_update({'_id': account['_id']}, {'$set': {'balances': new_balances}})
                    account_collection.find_one_and_update({'_id': to_account['_id']}, {'$set': {'balances.bank': new_to_balance}})
                    add_transaction(type='User Account to User Account', data={
                        'from': from_, 'to': to, 'amount': amount
                    })

                    return JSONResponse(
                        {
                            'message': 'successfully transferred funds', 'success': True,
                            'account': {'balances': new_balances}
                        }, status_code=200
                    )

                else:
                    return JSONResponse({'message': 'cant find to account', 'success': False}, status_code=404)

            elif transfer.type == 'user:coin-to-user:coin':
                from_ = transfer.data['from']
                to = transfer.data['to']
                amount = transfer.data['amount']
                from_bought_coin = bought_coin_collection.find_one({'owner': account['_id'], 'name': from_})
                to_bought_coin = bought_coin_collection.find_one({'wallet': to})

                if amount < 0:
                    return JSONResponse({'message': 'amount must be greater than 0', 'success': False}, status_code=400)

                if not from_bought_coin:
                    return JSONResponse({'message': 'cant find from bought coin', 'success': False}, status_code=404)

                if not to_bought_coin:
                    return JSONResponse({'message': 'cant find to bought coin', 'success': False}, status_code=404)

                from_coin = coin_collection.find_one({'name': from_bought_coin['name']})
                to_coin = coin_collection.find_one({'name': to_bought_coin['name']})

                if not from_coin:
                    return JSONResponse({'message': 'cant find from coin', 'success': False}, status_code=404)

                if not to_coin:
                    return JSONResponse({'message': 'cant find to coin', 'success': False}, status_code=404)

                if from_bought_coin['amount'] * from_coin['price'] < amount:
                    return JSONResponse({'message': 'insufficient funds', 'success': False}, status_code=400)

                from_diff = amount / from_coin['price']
                to_dif = amount / to_coin['price']

                new_from_amount = from_bought_coin['amount'] - from_diff
                new_to_amount = to_bought_coin['amount'] + to_dif

                bought_coin_collection.find_one_and_update({'_id': from_bought_coin['_id']}, {'$set': {'amount': new_from_amount}})
                bought_coin_collection.find_one_and_update({'_id': to_bought_coin['_id']}, {'$set': {'amount': new_to_amount}})
                add_transaction(type='User Coin to User Coin', data={
                    'from': from_, 'to': to, 'amount': amount
                })

                return JSONResponse(
                    {
                        'message': 'successfully transferred funds', 'success': True,
                        'account': {'balances': account['balances']}
                    }, status_code=200
                )

            else:
                return JSONResponse({'message': 'cant find transfer type', 'success': False}, status_code=404)
        else:
            return JSONResponse({'message': 'cant find account', 'success': False}, status_code=404)

    except jwt.exceptions.DecodeError:
        return JSONResponse({'message': 'invalid token', 'success': False}, status_code=401)

    except jwt.exceptions.ExpiredSignatureError:
        return JSONResponse({'message': 'token expired', 'success': False}, status_code=401)

    except Exception as e:
        return JSONResponse(
            {'message': 'unknown error', 'error': str(e), 'success': False}, status_code=500
        )